qiskit/qobj/_qobj.py

```diff
@@ -0,0 +1,209 @@
+        """
+        Return a valid representation of `obj` depending on its type.
+        """
+        if isinstance(obj, list):
+            return [cls._expand_item(item) for item in obj]
+        if isinstance(obj, QObjItem):
+            return obj.as_dict()
+        return obj
+
+    @classmethod
+    def from_dict(cls, obj):
+        """
+        Return a QObjItem from a dictionary recursively, checking for the
+        required attributes.
+
+        Returns:
+            QObjItem: a new QObjItem.
+
+        Raises:
+            QObjValidationError: if the dictionary does not contain the
+                required attributes for that class.
+        """
+        if not all(key in obj.keys() for key in cls.REQUIRED_ARGS):
+            raise QObjValidationError(
+                'The dict does not contain all required keys: missing "%s"' %
+                [key for key in cls.REQUIRED_ARGS if key not in obj.keys()])
+
+        return cls(**{key: cls._qobjectify_item(value)
+                      for key, value in obj.items()})
+
+    @classmethod
+    def _qobjectify_item(cls, obj):
+        """
+        Return a valid value for a QObjItem from a object.
+        """
+        if isinstance(obj, dict):
+            # TODO: should use the subclasses for finer control over the
+            # required arguments.
+            return QObjItem.from_dict(obj)
+        elif isinstance(obj, list):
+            return [cls._qobjectify_item(item) for item in obj]
+        return obj
+
+    def __reduce__(self):
+        """
+        Customize the reduction in order to allow serialization, as the QObjs
+        are automatically serialized due to the use of futures.
+        """
+        init_args = tuple(getattr(self, key) for key in self.REQUIRED_ARGS)
+        extra_args = {key: value for key, value in self.__dict__.items()
+                      if key not in self.REQUIRED_ARGS}
+        return self.__class__, init_args, extra_args
+
+
+class QObj(QObjItem):
+    """Representation of a QObj.
+
+    Attributes:
+        id (str): QObj identifier.
+        config (QObjConfig): config settings for the QObj.
+        circuits (list[QObjExperiment]): list of experiments.
+        type (str): experiment type (QASM/PULSE).
+        _version (str): QObj version.
+    """
+
+    REQUIRED_ARGS = ['id', 'config', 'circuits']
+
+    def __init__(self, id, config, circuits, **kwargs):
+        # pylint: disable=redefined-builtin,invalid-name
+        self.id = id
+        self.config = config
+        self.circuits = circuits
+
+        self.type = QObjType.QASM.value
+        self._version = QOBJ_VERSION
+
+        super().__init__(**kwargs)
+
+
+class QObjConfig(QObjItem):
+    """Configuration for a QObj.
+
+    Attributes:
+        max_credits (int): number of credits.
+        shots (int): number of shots.
+        backend_name (str): name of the backend.
+    """
+    REQUIRED_ARGS = ['max_credits', 'shots', 'backend_name']
```

Comment:
`max_credits` should be optional, not required.
Also, `backend_name` shouldn't belong in QobjConfig at all. Basically, a Qobj should not be tied to a backend, it should be possible to, for example, compile for a device, then run the output Qobj on a simulator.

Reply:
Hmm, I think this comment and the rest of comments about the classes in this particular file are part of the "to be done in future PRs"! I agree with your comments - but only as soon as we have something that resembles the final qobj schemas in place. At the moment, the class hierarchy and the required attributes, etc in this class have been defined following the _current output_ of compilation on purpose, in order to start with something compatible with the rest of the codebase, and work from there.